srtcore/core.cpp

```diff
@@ -9834,3 +10028,818 @@ bool CUDT::runAcceptHook(CUDT *acore, const sockaddr* peer, const CHandShake& hs
+        // is called, the only result of it would be that it will really get
+        // the very next sequence, even though this function doesn't know it
+        // yet, but surely in both cases the situation is the same: the medium
+        // is ready for reading, no matter what packet will turn out to be
+        // returned when reading is done.
+
+        ready = true;
+    }
+
+    // When the sequence number is behind the current one,
+    // stating that the readines wasn't checked otherwise, the reading
+    // function will not retrieve anything ready to read just by this premise.
+    // Even though this packet would have to be eventually extracted (and discarded).
+
+    if (ready)
+    {
+         m_pGlobal->m_EPoll.update_events(id(), m_sPollID, SRT_EPOLL_IN, true);
+    }
+}
+
+void CUDTGroup::updateWriteState()
+{
+    CGuard lg (m_GroupLock);
+    m_pGlobal->m_EPoll.update_events(id(), m_sPollID, SRT_EPOLL_OUT, true);
+}
+
+// The "app reader" version of the reading function.
+// This reads the packets from every socket treating them as independent
+// and prepared to work with the application. Then packets are sorted out
+// by getting the sequence number.
+int CUDTGroup::recv(char* buf, int len, SRT_MSGCTRL& w_mc)
+{
+    //// TEMPORARY STUB.
+    LOGC(dlog.Error, log << "CUDTGroup::recv: not implemented for type #" << m_type);
+    throw CUDTException(MJ_SETUP, MN_INVAL, 0);
+}
+
+CUDTGroup::ReadPos* CUDTGroup::checkPacketAhead()
+{
+    typedef map<SRTSOCKET, ReadPos>::iterator pit_t;
+    ReadPos* out = 0;
+
+    // This map no longer maps only ahead links.
+    // Here are all links, and whether ahead, it's defined by the sequence.
+    for (pit_t i = m_Positions.begin(); i != m_Positions.end(); ++i)
+    {
+        // i->first: socket ID
+        // i->second: ReadPos { sequence, packet }
+        // We are not interested with the socket ID because we
+        // aren't going to read from it - we have the packet already.
+        ReadPos& a = i->second;
+
+        int seqdiff = CSeqNo::seqcmp(a.sequence, m_RcvBaseSeqNo);
+        if ( seqdiff == 1)
+        {
+            // The very next packet. Return it.
+            m_RcvBaseSeqNo = a.sequence;
+            HLOGC(dlog.Debug, log << "group/recv: ahead delivery %"
+                    << a.sequence << "#" << a.mctrl.msgno << " from @" << i->first << ")");
+            out = &a;
+        }
+        else if (seqdiff < 1 && !a.packet.empty())
+        {
+            HLOGC(dlog.Debug, log << "group/recv: @" << i->first << " dropping collected ahead %"
+                    << a.sequence << "#" << a.mctrl.msgno << ")");
+            a.packet.clear();
+        }
+        // In case when it's >1, keep it in ahead
+    }
+
+    return out;
+}
+
+string CUDTGroup::StateStr(CUDTGroup::GroupState st)
+{
+    static string states [] = { "PENDING", "IDLE", "RUNNING", "BROKEN" };
+    if (int(st) < 5)
+        return states[st];
+    return string("UNKNOWN");
+}
```

Comment:
Isn't it better to return `const char*` to avoid copying?

Reply:
Yes, and potential crashes, too.

Reply:
You still return a string. instead of `const char*`

Reply:
UNKNOWN wasn't in the map - but still, should be possible...